prow/pod-utils/gcs/upload_test.go

```diff
@@ -95,8 +125,12 @@ func TestUploadToGcs(t *testing.T) {
 			t.Errorf("%s: expected an error but got none", testCase.name)
 		}
 
-		if count != (testCase.passingTargets + testCase.failingTargets) {
-			t.Errorf("%s: had %d passing and %d failing targets but only ran %d targets, not %d", testCase.name, testCase.passingTargets, testCase.failingTargets, count, testCase.passingTargets+testCase.failingTargets)
+		// the fail func will update count for each retry, while the flaky func only updates count on final try
+		finalFailingCount := testCase.failingTargets * uploadRetries
+		finalPassingCount := testCase.passingTargets + testCase.flakyTargets
+
+		if count != (finalPassingCount + finalFailingCount) {
```

Comment:
I think we want to validate both that we have the correct number of fails, successes and total

Reply:
IMO that is unnecessary, and just makes the test case more fragile than necessary.

What is important is that a) we have a flaky uploader that will not pass the first time b) that passes on subsequent attempts and c) this results in a successful upload.

Ideally we can tune whether we retry 2, 5, 50 times without rewriting the test case.

Reply:
Both tests are good, right? That we succeed if a retry succeeds _and_ that we're actually retrying the correct number of times. The former is a package-level test of exported behavior and the latter is a unit test of package-local behavior.

Reply:
I do not consider both tests good, no. Validating that we retry exactly 5 times instead of 2 or 6 is an over-specified test in this scenario.

There's nothing important about 5, just an arbitrary guesstimate. Empirical data may cause us to feel differently, and changing it should not require modifying the test.

(as a convenience I'd be likely to set a `const retryCount = 5` and use this for both the flaky test func and the retry impl loop, for the same reason: this allows me to later do `retryCount = 10` or `retryCount = 2` without risking regressing the test case, but that is different from validating the number of retries).